### PR TITLE
Normalize the total energy by natoms

### PR DESCRIPTION
**Pull request by nw13slx**
## Description
allow to use energy/atom for loss function

## Motivation and Context
could be useful for datasets with varying numbers of atoms

## How Has This Been Tested?
unit tests added to tests/unit/train/test_loss module

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [x] My code follows the code style of this project and has been formatted using `black`.
- [x] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [x] `example.yaml` (and other relevant `configs/`) have been updated with new or changed options.
- [x] I have updated `CHANGELOG.md`.

### CODE REVIEW

File: configs/minimal_eng.yaml [hunk 1]
```diff
@@ -34,9 +33,35 @@ n_train: 5
 n_val: 5
 batch_size: 1
 max_epochs: 10
+learning_rate: 0.0
```
Comment (Linux-cpp-lisp) [inline configs/minimal_eng.yaml:36]: ?

File: configs/minimal_eng.yaml [hunk 2]
```diff
@@ -34,9 +33,35 @@ n_train: 5
...
+      report_per_component: False
+  - - total_energy
+    - mae
+    - PerAtom: True                        # if true, per species contribution is counted separately
+  - - total_energy
+    - mae
+    - PerAtom: False                       # if true, per species contribution is counted separately
+  - - total_energy
+    - rmse
+    - PerAtom: True                        # if true, per species contribution is counted separately
```
Comment (Linux-cpp-lisp) [inline configs/minimal_eng.yaml:64]: "per species contribution is counted separately" means?

File: nequip/train/_loss.py [hunk 3]
```diff
@@ -61,6 +62,40 @@ def __call__(
...
+class PerAtomLoss(SimpleLoss):
+    def __call__(
+        self,
+        pred: dict,
+        ref: dict,
+        key: str,
+        mean: bool = True,
+    ):
+        # zero the nan entries
+        has_nan = self.ignore_nan and torch.isnan(ref[key].mean())
```
Comment (Linux-cpp-lisp) [inline nequip/train/_loss.py:74]:
```suggestion
        has_nan = self.ignore_nan and torch.isnan(ref[key].sum())
```
might be cheaper?

File: nequip/train/metrics.py [hunk 4]
```diff
@@ -53,34 +56,40 @@ def __init__(
...
         self.kwargs = {}
         for component in components:
 
             key, reduction, params = Metrics.parse(component)
 
-            functional = params.pop("functional", "L1Loss")
+            params["PerSpecies"] = params.get("PerSpecies", False)
+            params["PerAtom"] = params.get("PerAtom", False)
+
+            param_hash = Metrics.hash_component(component)
```
Comment (Linux-cpp-lisp) [inline nequip/train/metrics.py:69]: why is this necessary instead of the reduction name?

Or just making it a list and indexing it in order?

File: nequip/train/metrics.py [hunk 5]
```diff
@@ -184,16 +212,22 @@ def flatten_metrics(self, metrics, type_names=None):
...
 
-            stat = self.running_stats[key][reduction]
-            per_species = self.per_species[key][reduction]
+            stat = self.running_stats[key][param_hash]
+            per_species = params["PerSpecies"]
 
             if per_species:
                 if stat.output_dim == tuple():
+                    if type_names is None:
+                        type_names = [i for i in range(len(value))]
```
Comment (Linux-cpp-lisp) [inline nequip/train/metrics.py:230]: is this a backwards compatability hack?

File: configs/minimal_eng.yaml (hunk 1 above)
Comment (nw13slx) [reply configs/minimal_eng.yaml:36]: ??!! I don't know how that happened... 🤦‍♀️ I'll check out the original version

File: configs/minimal_eng.yaml (hunk 2 above)
Comment (nw13slx) [reply configs/minimal_eng.yaml:64]: will fixed

File: nequip/train/metrics.py (hunk 4 above)
Comment (nw13slx) [reply nequip/train/metrics.py:69]: it turns out that I want to have the same reduction, with and without the "per atom" flag. And in case we have more different things in the future, it's easier to just hash the whole dict.

File: nequip/train/metrics.py (hunk 5 above)
Comment (nw13slx) [reply nequip/train/metrics.py:230]: yeah. I kept running in to the issue where `type_name` is none for `nequip-train minimal_eng.yaml`... Still don't know where exactly the bug is. But if we don't allow `type_name` to be none, we shouldn't let it to be an optional argument in the first place.

File: configs/minimal_eng.yaml (hunk 1 above)
Comment (nw13slx) [inline configs/minimal_eng.yaml:36]:
```suggestion
```

File: nequip/train/metrics.py (hunk 5 above)
Comment (Linux-cpp-lisp) [reply nequip/train/metrics.py:230]: ... it should never be none?

**Comment by whu-dft** [comment]
Hi, nequip is a very nice work! I currently study the nequip codes but was puzzled by the perAtomLoss as follows:

image

You devide the total_energy by N (number of atoms in the molecule) twice. For the first time, it's easy to understand that the perAtomLoss is calculated with total energy divided by N atoms. However, **I was puzzled when the func_name equals to "MSELoss", you divided the loss by N again**.

Let me make things cleaner, I think the perAtomLoss should be calculated in the case of MSELoss (one batch samples) by

```python
loss = (pred['total_energy'] - ref['total_energy'])**2 / N    # per atom energy loss
loss = loss.mean()  # average over batch samples
```

The number of atoms N is divided  **only once**. 

Sorry to interrup you, I hope I've made a mistake.

**Comment by Linux-cpp-lisp** [comment]
Hi @whu-dft,

Thanks for your interest in our work!

This is intentional since when we are using an MSELoss `loss` has the **square** of the error, so we want to divide by the $N$ twice, which is the same as dividing by $N^2$, which is the same as dividing by $N$ inside of the $( \cdot )^2$ to give the expected expression for the per-atom loss (which you can see written out in the appendix of the Allegro paper, for example).

**Comment by whu-dft** [comment]
Oh, I see. I'm so glad to get your response soon. Thank you very much!  ^_^
